doc/src/explanation/solve_output.rst

```diff
@@ -0,0 +1,106 @@
+interaction rather than programmatic use. The type of output will depend on the
+type of equation(s) (and how they are entered) and the number of symbols that
+are provided (and how they are provided).
+
+Since this output is controllable by setting `dict` or `set` to True
+(when not dealing with relational expressions) the following is an explanation for
+those that might prefer the *default* output but wish to understand better the
+conditions under which it is given.
+
+    >>> from sympy import sqrt, exp, solve, Symbol, Eq
+    >>> from sympy.abc import x, y, a, b
+
+1. An empty list indicates that no solution was found.
+
+    >>> [] == solve(sqrt(x) + 1) == solve(sqrt(x) + 1, dict=True)
+    True
+    >>> solve(sqrt(x) + 1, set=True)
+    ([x], set())
+
+2. A list of values is given when the symbol to solve for was
+unambiguous in context because a) the equation was univariate and
+no symbols were specified or b) a single symbol was specified as
+being of interest.
+
+    >>> solve(x**2 - 4)
+    [-2, 2]
+    >>> solve(x - y - 1, x)
+    [y + 1]
+
+3. A single dictionary with keys being symbols and values being the solutions
+for those symbols is the result when equations are passed as a list and are
+all linear in the symbols given. Such a system is automatically generated
+for a single equation (not passed as a list) if there is an undetermined
+coefficients solution for the symbols specified.
+
+    >>> solve([x + y - 2, x - y + 2], x, y)
+    {x: 0, y: 2}
+    >>> solve(a*x - 2*x + b - 5, {a, b})  # undetermined coefficients
+    {a: 2, b: 5}
+
+4. A list of tuples, each giving a solution for the symbols in the order
+they were given, is given when a list of equations containing at least
+one nonlinear equation or a list of symbols is given in a well defined
+order. (This is also the format for the tuples in the set returned when
+using the flag ``set=True``.)
+
+    >>> solve(x - 1, x, y)  # more than one symbol
+    [(1, y)]
+    >>> solve([x**2], x)  # list with nonlinear equation
+    [(0,)]
+    >>> solve([x**2 - 1], x)
+    [(-1,), (1,)]
+    >>> solve([x**2 - y, x - 3], x, y)  # nonlinear and multiple symbols
+    [(3, 9)]
+
+5. A list of dictionaries is returned when the expression was not univariate or
+there was a nonlinear expression in a list *and* the order of
+symbols would otherwise be ambiguous because a) no symbols were passed or b) the
+symbols were passed as a set. The dictionary will only contain values that are
+distinct from the keys. (This is also the format selected with ``dict=True``.)
+
+    >>> solve(x - y)
+    [{x: y}]
+    >>> solve([exp(x) - 1, x**2])
+    [{x: 0}]
+    >>> solve([x + y - 2, x**2 - y + 2], {x, y})
+    [{x: -1, y: 3}, {x: 0, y: 2}]
+
+6. A boolean expression is returned when a relational expression other
+than an Equality is given as an expression to solve. A single Equality
+or a more complicated relational expression might be returned.
```

Comment:
It would be good to clarify here that inequalities cannot be used to filter the output of a system of equations. I think that's what most people want solve to do but it doesn't work that way.

Really we should recommend to use `reduce_inequalities` rather than using `solve` in this way.